### PR TITLE
moco: deploy moco v0.8.2

### PR DESCRIPTION
**Pull request by yokaze**
Signed-off-by: Daichi Sakaue <daichi-sakaue@cybozu.co.jp>

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -146,11 +146,8 @@ update-metallb:
 .PHONY: update-moco
 update-moco:
 	$(call get-latest-gh,cybozu-go/moco)
-	rm -rf /tmp/moco
-	cd /tmp; git clone --depth 1 -b $(latest_gh) https://github.com/cybozu-go/moco
-	rm -rf moco/base/upstream/*
-	cp -r /tmp/moco/config/* moco/base/upstream
-	rm -rf /tmp/moco
+	curl -sfL -o moco/base/upstream/moco.yaml \
+		https://github.com/cybozu-go/moco/releases/latest/download/moco.yaml
```
Comment (ymmt2005) [inline Makefile:150]: Download a tagged manifest.

File: Makefile [hunk 2]
```diff
@@ -146,11 +146,8 @@ update-metallb:
...
 	sed -i -E 's/newTag:.*$$/newTag: $(patsubst v%,%,$(latest_gh))/' moco/base/kustomization.yaml
```
Comment (ymmt2005) [inline Makefile:151]: We don't need to replace the image as MOCO is our product.

File: moco/base/kustomization.yaml [hunk 3]
```diff
@@ -0,0 +1,5 @@
+apiVersion: kustomize.config.k8s.io/v1beta1
```
Comment (ymmt2005) [inline moco/base/kustomization.yaml:1]: Why don't we use the upstream as is?

File: test/testdata/moco.yaml [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+        image: quay.io/cybozu/moco-mysql:8.0.24
+        resources:
+          requests:
+            memory: "512Mi"
+  volumeClaimTemplates:
+  # At least a PVC named "mysql-data" must be defined.
+  - metadata:
+      name: mysql-data
+    spec:
+      accessModes: [ "ReadWriteOnce" ]
```
Comment (ymmt2005) [inline test/testdata/moco.yaml:38]: Use topolvm-provisioner.

File: test/testdata/moco.yaml [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+                values:
+                - test
+            topologyKey: "kubernetes.io/hostname"
+      containers:
+      # At least a container named "mysqld" must be defined.
+      - name: mysqld
+        image: quay.io/cybozu/moco-mysql:8.0.24
+        resources:
+          requests:
+            memory: "512Mi"
```
Comment (ymmt2005) [inline test/testdata/moco.yaml:32]: For tests, we don't need to give resource requests.

File: test/testdata/moco.yaml [hunk 6]
```diff
@@ -0,0 +1,41 @@
+# This manifest is based on the this example (https://github.com/cybozu-go/moco/blob/v0.8.1/docs/usage.md).
+apiVersion: moco.cybozu.com/v1beta1
+kind: MySQLCluster
+metadata:
+  namespace: test-moco
```
Comment (ymmt2005) [inline test/testdata/moco.yaml:5]:
```suggestion
  namespace: sandbox
```

File: test/moco_test.go [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+	appsv1 "k8s.io/api/apps/v1"
+)
+
+//go:embed testdata/moco.yaml
+var mocoYAML []byte
+
+func prepareMoco() {
+	It("should deploy mysqlcluster", func() {
+		By("creating mysqlcluster")
+		createNamespaceIfNotExists("test-moco", false)
```
Comment (ymmt2005) [inline test/moco_test.go:20]: Let's use `sandbox`.

File: test/moco_test.go [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+				return fmt.Errorf("AvailableReplicas is not 1: %d", int(deployment.Status.AvailableReplicas))
+			}
+			return nil
+		}).Should(Succeed())
+	})
+
+	It("should work", func() {
+		By("waiting mysqlcluster is ready")
+		Eventually(func() error {
+			stdout, _, err := ExecAt(boot0, "kubectl", "--namespace=test-moco", "get", "mysqlcluster/test", "-o", "jsonpath='{.status.ready}'")
```
Comment (ymmt2005) [inline test/moco_test.go:49]: Does `status.ready` really exist?
I don't think I implemented such a thing.

File: test/moco_test.go [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+			}
+
+			if string(stdout) != "True" {
+				return errors.New("MySQLCluster is not ready")
+			}
+			return nil
+		}).Should(Succeed())
+
+		By("running kubectl moco mysql")
+		stdout, stderr, err := ExecAt(boot0, "kubectl", "moco", "-n", "test-moco", "mysql", "-u", "root", "my-cluster", "--", "--version")
```
Comment (ymmt2005) [inline test/moco_test.go:61]: I don't think `root` user exists.  Please check the manual.

File: test/moco_test.go (hunk 8 above)
Comment (yokaze) [reply test/moco_test.go:49]: Currently the test code is same as that for MOCO v0.7.0, I'm just seeing the CI result and making fixes.

Comment (ymmt2005) [reply test/moco_test.go:49]: Use this.

```console
$ kubectl -n foo get mysqlcluster/single -o jsonpath='{.status.conditions[?(@.type=="Healthy")].status}'
True
```

**Comment by yokaze** [comment]
This PR should be merged with squash.

File: test/testdata/moco.yaml [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+kind: MySQLCluster
+metadata:
+  namespace: sandbox
+  name: test
+spec:
+  # replicas is the number of mysqld Pods.  The default is 1.
+  replicas: 1
+  podTemplate:
+    spec:
+      affinity:
```
Comment (ymmt2005) [inline test/testdata/moco.yaml:12]: This cluster consists of 1 pod, so there is no need for (anti-)affinities.

File: test/moco_test.go [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+				return err
+			}
+			deployment := new(appsv1.Deployment)
+			err = json.Unmarshal(stdout, deployment)
+			if err != nil {
+				return err
+			}
+
+			if int(deployment.Status.AvailableReplicas) != 2 {
+				return fmt.Errorf("AvailableReplicas is not 2: %d", int(deployment.Status.AvailableReplicas))
```
Comment (ymmt2005) [inline test/moco_test.go:40]: int cast is not needed.

File: test/moco_test.go [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+			}
+			return nil
+		}).Should(Succeed())
+	})
+
+	It("should work", func() {
+		By("waiting mysqlcluster is ready")
+		Eventually(func() error {
+			stdout, _, err := ExecAt(boot0, "kubectl", "--namespace=sandbox", "get", "mysqlcluster/test", "-o", "{.status.conditions[?(@.type=='Healthy')].status}")
```
Comment (ymmt2005) [inline test/moco_test.go:49]:
```suggestion
			stdout, _, err := ExecAt(boot0, "kubectl", "--namespace=sandbox", "get", "mysqlcluster/test", "--template", "{.status.conditions[?(@.type=='Healthy')].status}")
```

**Review by ymmt2005 — CHANGES_REQUESTED** [review]
Please fix.

**Pull request merged**